third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1184 @@
+                        .collect(),
+                )
+            })
+            .collect();
+
+    // Calculate the list of inline functions which call at least one other inline function.
+    let inline_functions_calling_others: Vec<QualifiedFunId> = inline_functions_with_callees
+        .iter()
+        .filter(|(_, callees)| !callees.is_empty())
+        .map(|(caller_fnid, _)| caller_fnid)
+        .cloned()
+        .collect();
+
+    // Check for cycles
+    let cycles = check_for_cycles(&inline_functions_with_callees);
+    if !cycles.is_empty() {
+        for cycle in cycles {
+            let start_fnid = cycle.first().unwrap();
+            let func_env = env.get_function(*start_fnid);
+            let path_string: String = cycle
+                .iter()
+                .map(|fnid| env.get_function(*fnid).get_name_str())
+                .collect::<Vec<String>>()
+                .join(" -> ");
+            let mut call_details: Vec<_> = cycle
+                .iter()
+                .zip(cycle.iter().skip(1).chain([*start_fnid].iter()))
+                .flat_map(|(f, g)| {
+                    let sites_ids = call_site_locations.get(&(*f, *g)).unwrap();
+                    let f_str = env.get_function(*f).get_name_str();
+                    let g_str = env.get_function(*g).get_name_str();
+                    let msg = format!("call from {} to {}", f_str, g_str);
+                    sites_ids
+                        .iter()
+                        .map(move |node_id| (env.get_node_loc(*node_id), msg.clone()))
+                })
+                .collect();
+            let msg = format!(
+                "recursion during function inlining not allowed: {} -> {}",
+                path_string,
+                func_env.get_name_str()
+            );
+            let loc = call_details.first_mut().unwrap().0.clone();
+            env.diag_with_labels(Severity::Error, &loc, &msg, call_details);
+        }
+        return Err(());
+    }
+
+    // Compute post-order of inlined_functions which call others.
+    let po_inline_functions = postorder(
+        &inline_functions_calling_others,
+        &inline_functions_with_callees,
+    );
+
+    // Identify subset of non-inline functions which call inline functions.  Order doesn't matter here.
+    let non_inline_functions_needing_inlining: Vec<QualifiedFunId> = function_callees
+        .iter()
+        .filter(|(fnid, callees)| {
+            !env.get_function(**fnid).is_inline()
+                && callees
+                    .iter()
+                    .any(|fnid2| env.get_function(*fnid2).is_inline())
+        })
+        .map(|(fnid, _callees)| fnid)
+        .cloned()
+        .collect();
+
+    let result: Vec<QualifiedFunId> =
+        chain(po_inline_functions, non_inline_functions_needing_inlining).collect();
+    Ok(result)
+}
+
+// Calculate a bottom-up traversal for entries, given the provided callee map.
+fn postorder<T: Ord + Copy + Debug>(
```

Comment:
Because this is isolated, please consider adding unit test at the bottom of the file:

```
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_postorder() {
        let entries = vec![1, 2, 3, 4, 5, 7];
        let call_graph = BTreeMap::from([
            (1, BTreeSet::from([2, 3])),
            (2, BTreeSet::from([4])),
            (3, BTreeSet::from([4])),
            (4, BTreeSet::from([5, 6])),
            (5, BTreeSet::from([7])),
            (6, BTreeSet::new()),
            (7, BTreeSet::from([8])),
            (9, BTreeSet::new()),
        ]);
        assert!(postorder(&entries, &call_graph) == vec![7, 5, 6, 4, 3, 2, 1]);
    }
}
```

Reply:
Great, thanks!

Reply:
Although, your assertion is wrong.  I've correct it in the code.